### PR TITLE
change default prefix

### PR DESCRIPTION
**Pull request by Baaaaam**
this change the default install prefix for symmetric (on MACOS -- no change on ubuntu)

the standard `--prefix` behavior is still available...

### CODE REVIEW

**Comment by Baaaaam** [comment]
@FlanFlanagan this is the other PR, which set the default user install path to `~/.local/lib/python...` for Cymetric python lib

File: setup.py [hunk 1]
```diff
@@ -37,13 +37,19 @@
...
 
 
 VERSION = '1.5.0'
 IS_NT = os.name == 'nt'
 
+HOME = os.getenv("HOME")
+PY_VERSION = str(sys.version_info[0]) +'.' + str(sys.version_info[1])
+
+site.USER_BASE = HOME + '/.local'
+site.USER_SITE = HOME + '/.local/lib/python'+ PY_VERSION + '/site-packages'
```
Comment (gonuke) [inline setup.py]: Change the default location to match this.

**Comment by Baaaaam** [comment]
@gonuke I updated the change the installation PREFIX:
now it check if --prefix is an argument if not it add a `--prefix=~/.local` (and remove the `--user` if one)
I was scared that playing with the default Python path may prevent python to find already installed package...

this forces the installation to be in .local/ unless --prefix=XXX is provided

File: setup.py [hunk 2]
```diff
@@ -46,6 +47,18 @@
 
 
 def main():
+    
+    arguments = []
+    for arg in sys.argv[1:]:
+        if not arg == "--user":
+            arguments.append(arg)
```
Comment (gonuke) [inline setup.py]: Should we add a message to the user that we are overriding this?

File: setup.py [hunk 3]
```diff
@@ -46,6 +47,18 @@
...
+   
+    if not( any("--prefix" in arg for arg in arguments)):
+        home = os.getenv("HOME")
+        prefix = '--prefix=' + home + '/.local'
```
Comment (gonuke) [inline setup.py]: add a message to the user that we are overriding the default?

**Comment by gonuke** [comment]
Works for me (at least as much as it ever does)

**Comment by gonuke** [comment]
Just a little more user feedback and I think it's OK

**Comment by Baaaaam** [comment]
this now shows:
```

WARNING: Default installation path have been overrideto'~/.local'.
WARNING: To specify the install prefix use--prefix=your/install/path flag.

```

or 

```

WARING: --user flag have been desabled.
WARNING: To specify the install prefix use--prefix=your/install/path flag.

```

if respectively not flag or `--user flag`  is provided

I put 1 empty  line after and before to provide visibility to the msg.

**Comment by FlanFlanagan** [comment]
This installs for me on ubuntu 16.04. +1

File: setup.py [hunk 4]
```diff
@@ -46,6 +46,26 @@
...
+            arguments.append(arg)
+
+    if not( any("--prefix" in arg for arg in arguments)):
+        home = os.getenv("HOME")
+        prefix = '--prefix=' + home + '/.local'
+        arguments.append(prefix)
+
+        if any("--user" in arg for arg in sys.argv):
+            print("\nWARNING: --user flag have been disabled and installation"
+                  " path set to '~/.local'.")
```
Comment (gonuke) [inline setup.py]: add a reason "to be consistent with the Cyclus installation options."

**Comment by Baaaaam** [comment]
@gonuke is it better ?

**Comment by Baaaaam** [comment]
@gonuke Shall I add a verification if both --user and --prefix are provided ?

**Comment by gonuke** [comment]
We can skip a test for --user and --prefix being given simultaneously since it's reasonable to assume that the latter would override the former anyway.

**Comment by gonuke** [comment]
Thanks @Baaaaam

**Comment by scopatz** [comment]
I have to say I am -1 on this. We shouldn't break comparability with all other setup.pys in the world. Also, this only seems to break it such that it just reimplements the same semantic meaning as what it was doing before. I am sorry I missed it.

Can you explain please what was wrong? I am not sure I fully understand the problem since I am not a mac user.  Even if you want to set a different default prefix, there is no reason to touch user.

I am happy to help work towards a better solution, but I don't think that this is the right one.

**Comment by Baaaaam** [comment]
@scopatz 
on mac the default Python path is `~/Library/Python/...` the problem is the Cyclus lib was install by default in `~/.local/lib/pythonXY`
this allow both to be installed in the same place

there is a more detailed explanation in cyclus/cyclus#1324

**Comment by scopatz** [comment]
Hi @Baaaaam, I guess I still don't understand why breaking the default install location is preferable. It seems like it is asking for problems, like making the users set their Python path when they wouldn't normally need to. 

Also, this definitely seems like a Mac-only issue. So the fix shouldn't affect Linux at all, in my opinion.

**Comment by scopatz** [comment]
In my experience, it generally is more trouble than it is worth to try to beak the defaults, unless you are going to do so in the context of a larger packaging system.

**Comment by scopatz** [comment]
Why not tell user's to just install with something like, 

```console
$ ./setup.py install --prefix $(cyclus --install-path)
```

**Comment by Baaaaam** [comment]
we had 2 possible changes in mind:
- change the way cyclus was installing by default the cyclus python lib (it forces to have the same prefix than cyclus, even if the prefix was not user specified)
- change the way cymetric was installing its library.

we decided to change it this way to have all the cyclus "stuff" in the same place, and to have a behavior of the symmetric installation script similar to the cyclus one: 
- default installation in ~/.local/lib/
- no use of --user,
- use --prefix to manually specify the prefix

**Comment by Baaaaam** [comment]
I think we decided to have the similar way to install Cymetric and Cyclus.

We can definitely revert that, but if so Cyclus will have an issue: the cyclus python lib is installed in ~/.local/lib by default, which is not the default python prefix on MacOS.

see cyclus/cyclus#1328 PR which do this.

We wanted the release out, so we had to decide something. This could be reverted, and a PR is almost ready to fix cyclus in the same way.

**Comment by scopatz** [comment]
> change the way cyclus was installing by default the cyclus python lib

Right because the cyclus cmake scripts should fully respect the install location that the user provides at cyclus install time. 

> we decided to change it this way to have all the cyclus "stuff" in the same place

Well, it shouldn't matter that they are installed in the same prefix location. Python should be able to find the different packages just fine. So I guess I still don't understand why it is important to enforce this. 

> default installation in ~/.local/lib/
> no use of --user,
> use --prefix to manually specify the prefix

What is important is that, like with cyclus, cymetric fully respects the specified install location provided by the user. This PR breaks the default location and the `--user` install location. These and `--prefix` are strongly expected to go to the place that they always go to.  

If we want the user to install to a different location, we should ask them to in the readme (with the command above), but not force them to. 

In summary, I don't think that either this change or the proposed change to cyclus are correct.  Rather this should be a one-line change to the readme.

**Comment by scopatz** [comment]
I have a potential fix in #115

**Comment by Baaaaam** [comment]
thx @scopatz for this fix, I will look at it.

> > change the way cyclus was installing by default the cyclus python lib
> Right because the cyclus cmake scripts should fully respect the install location that the user provides at cyclus install time.

Actually by default the installation of the Cyclus python lib are done using: `--prefix=${CMAKE_INSTALL_PREFIX}` where CMAKE_INSTALL_PREFIX is set by default to `~/.local/lib` (which is not the default python lib on MacOS), even if the user don't specify any $PREFIX.  (see install.py `l123  localdir = absexpanduser('~/.local')` )

I have proposed a a fix for this: cyclus/cyclus#1328. This use the --user flag for the installation of the Cyclus python library when user don't specify explicitly the `--prefix`.
this has been closed, but feel free to re-open and review it.

**Comment by scopatz** [comment]
> thx @scopatz for this fix, I will look at it.

Thanks!

> Actually by default the installation of the Cyclus python lib are done using: --prefix=${CMAKE_INSTALL_PREFIX}

Yeah I know. That is to make the cyclus install respect the prefix the user set via the standard mechanism by which CMake specifies the install location. I have reviewed  cyclus/cyclus#1328, and the problem with it (as I said above) is that it breaks the cmake default and does not respect the user setting.

**Comment by Baaaaam** [comment]
Oh sorry I missed that point, I think I know understand what your are saying. 

But is it not strange to have both default installation (Cyclus and Cymetric) not installing python lib in the same place ? (Even if it is in the same place when following the recommended instruction)

**Comment by scopatz** [comment]
Sure, it is a bit odd. But strange is better than broken.  Changing defaults like this almost always breaks things, in the short term or long term, for users or packagers or developers. And problems caused by these kinds of changes are incredibly hard to track down and debug. I spent the whole spring and summer tracking down these kinds of bugs this because our build system was so broken. 

I think people have to be able to read the installation docs or be responsible for doing something different. However, having broken builds and incomplete is solely our responsibility as developers.
